smarts/core/controllers/imitation_controller.py

```diff
@@ -32,43 +37,51 @@ def perform_action(cls, dt, vehicle, action):
         if isinstance(action, (int, float)):
             # special case:  setting the initial speed
             if isinstance(chassis, BoxChassis):
-                vehicle.control(vehicle.pose, action)
+                vehicle.control(vehicle.pose, action, dt)
             elif isinstance(chassis, AckermannChassis):
                 chassis.speed = action  # hack that calls control internally
             return
 
         assert isinstance(action, (list, tuple)) and len(action) == 2
-        # acceleration in m/s^2, angluar_velocity in rad/s
+        # acceleration is vector in m/s^2, angular_velocity is scalar in rad/s
         acceleration, angular_velocity = action
+        target_heading = (vehicle.heading + angular_velocity * dt) % (2 * math.pi)
 
         if isinstance(chassis, BoxChassis):
-            target_speed = vehicle.speed + acceleration * dt
-            target_heading = (vehicle.heading + angular_velocity * dt) % (2 * math.pi)
-            speed_x, speed_y = radians_to_vec(target_heading)
-            new_position = [
-                vehicle.position[0] + dt * speed_x * target_speed,
-                vehicle.position[1] + dt * speed_y * target_speed,
-                vehicle.position[2],
-            ]
-            new_pose = replace(
-                vehicle.pose,
+            speed_x, speed_y = radians_to_vec(vehicle.heading) * vehicle.speed
+            speed_x += acceleration[0] * dt
+            speed_y += acceleration[1] * dt
+            new_position = np.array(
+                (
+                    vehicle.position[0] + dt * speed_x,
+                    vehicle.position[1] + dt * speed_y,
+                    vehicle.position[2],
+                )
+            )
+            new_pose = Pose(
                 orientation=fast_quaternion_from_angle(target_heading),
                 position=new_position,
             )
-            vehicle.control(new_pose, target_speed)
+            vehicle.control(new_pose, np.linalg.norm((speed_x, speed_y)), dt)
```

Comment:
I do not see a reason for normalising the speed here. Shouldn't it just be the speed?

Reply:
@Gamenot : I think this is correct since `np.linalg.norm` is representing the norm of speed vector not its normalization

Reply:
This is just a way to get back to a scalar speed from the two vector components.  We're just using the magnitude of the vector.  But we haven't yet computed this new scalar value yet in this function.   (Note that `speed_x` and `speed_y` have been updated using the acceleration so that this will be a different value now than `vehicle.speed` and there is no more `target_speed` variable.)

Reply:
Oh, sorry, I keep forgetting that `np.linalg.norm` is for scalar. :facepalm: